Repo Gardening: add logic to get label suggestions from OpenAI

jeherve opened the pull request:
## Proposed changes:

This will be triggered as part of the existing "Triage issues" task, only when an issue is opened.

It will fetch the list of "Feature" and "Feature Group" labels available in the repo, and then send the issue contents to OpenAI alongside with that list of labels.
OpenAI will respond with a list of labels that could match the issue.
We then apply those labels to the issue.

> [!NOTE]
> In order to help us test this change with real life data, I added extra logic to this PR. The automated labeling will now only happen on issues labeled with the "[Experiment] Automated labeling" label. Only when that label is present will we ask OpenAI for suggestions. In addition, and again to facilitate testing, the bot will also comment on the issue to say what was suggested and why. I'm hoping it can help us iterate on the prompt faster.

**To do**

- [x] Figure out OpenAI authentication: PCYsg-Vp7-p2
- [x] Include issue title in info sent to OpenAI

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* Internal reference: pdqkMK-1kt-p2

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

This can only be tested in a fork of this repo so far, where you will have to push commits from this branch to trunk. There, add a new `OPENAI_API_KEY` secret with the key from the secret store, secret id 12313.

* Open a new issue.
* Label it with "[Experiment] Automated labeling"
* Let the action run (you can check the Actions tab in your repo to follow progress).
* You should see something like this (labels added, comment posted, label indicated issue has been processed): https://github.com/jeherve/jetpack/issues/152

jeherve committed:
Repo Gardening: add logic to get label suggestions from OpenAI

This will be triggered as part of the existing "Triage issues" task, only when an issue is opened.

It will fetch the list of "Feature" and "Feature Group" labels available in the repo, and then send the issue contents to OpenAI alongside with that list of labels.
OpenAI will respond with a list of labels that could match the issue.
We then apply those labels to the issue.

Internal reference: pdqkMK-1kt-p2

jeherve committed:
Fix output format and error handling

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/openai-auto-labeling-logic

jeherve committed:
Clarify usage in readmes.

jeherve committed:
Include issue title in text sent to OpenAI

jeherve committed:
Update prompt

jeherve commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +13,81 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+Here are the existing labels and their descriptions:
+${ repoLabels.join( ', ' ) }
+
+Analyze the issue and suggest relevant labels. Rules:
+- Don't search for and add more labels if the existing labels match your initial assessment.
+- Use only existing labels provided.
+- Include 1 '[Feature Group]' label.
+- Include 1 to 3 '[Feature]' labels.
+- Briefly explain each label choice in 1 sentence.
+- Format your response as a JSON object with 'labels' and 'explanations' keys.
```
@inaikem I tweaked your prompt a bit. I removed the "Product" and "Type" rules since those are currently not pulled in our label sync, not used in every repo, and in the case of "Type", they should already be added by the issue template. Happy to change that if you'd like though.

dereksmart replied:
Consider using OpenAI's [Structured Outputs](https://platform.openai.com/docs/guides/structured-outputs/introduction) to standardize the json response. It touts a 100% accuracy in the response format.

jeherve replied:
While Structures Outputs may not be available for our model yet, I went with JSON Mode which seems like a good option too: c23f7ff944812b5724f525971b5b2e0ff6486fda

jeherve committed:
Trigger the auto-labeling only on issues with a specific label

jeherve committed:
During testing, comment to the issue explaining the label choices

anomiex commented on projects/github-actions/repo-gardening/action.yml:
```diff
@@ -56,6 +56,10 @@ inputs:
     required: false
     default: ""
+  openai_api_key:
+    description: "OpenAI API"
```
```suggestion
    description: "OpenAI API key"
```

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+
+Example response format:
+{
+    "labels": ["[Product] WordPress.com", "[Feature Group] User Interaction & Engagement", "[Feature] Comments"],
```
Not sure if this might confuse the AI when it was told not to include "[Product]" labels and when no explanation for the label is included.
```suggestion
    "labels": ["[Feature Group] User Interaction & Engagement", "[Feature] Comments"],
```

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+	try {
+		parsedResponse = JSON.parse( response );
+	} catch ( error ) {
+		debug(
+			`triage-issues: OpenAI did not send back the expected JSON-formatted response. Error: ${ error }`
+		);
+		return suggestions;
```
I wonder if it would be useful to also say what OpenAI did return, rather than just the error message?

Maybe just move line 80 up and directly print the response instead of re-stringifying the parsed response?

jeherve replied:
👍 Done in e339f62dd10f67906a53795a149af965eead692c

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+
+	// Display the explanations in the action logs.
+	const explanationString = Object.entries( explanations ).join( ' // ' );
+	debug( `triage-issues: OpenAI suggested the following labels: ${ explanationString }` );
```
This seems somewhat redundant to the printing of the whole response, if someone cares couldn't they read it out of the JSON directly?

Also, this is assuming the AI didn't screw up and omit "explanation" (so it's undefined) or set it to a non-object. Or set different labels in "explanation" and "labels".

jeherve replied:
👍 Done in 0b3744450b5c645956edb53fce773cf8097a9976

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+        "[Feature Group] User Interaction & Engagement": "The issue involves how users interact with the platform.",
+        "[Feature] Comments": "Specifically, it's about the commenting functionality."
+    }
+}`;
```
I wonder if it would be more straightforward to have the AI respond with a single object, something like this
```json
{
    "[Feature Group] User Interaction & Engagement": "The issue involves how users interact with the platform.",
    "[Feature] Comments": "Specifically, it's about the commenting functionality."
}
```
and use `Object.keys()` to get the labels. Less opportunity for it to get confused and mismatch the two lists. Or does doing it like this make it more likely to give useful answers?

jeherve replied:
We can certainly give it a try. a8a5bd74ed6dc7fee9a7df86e136068955574cf7

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+		return true;
+	}
+
+	// Next, check if the current event was a [Experiment] Automated labeling label being added.
+	if (
+		'labeled' === action &&
+		eventLabel.name &&
+		'[Experiment] Automated labeling' === eventLabel.name
+	) {
+		return true;
+	}
```
Is there a situation where `getLabels()` would not return the experiment label but this check on the action would still pass?

Or, on the other hand, if it works to test this directly, why do we need to do `getLabels()` at all?

jeherve replied:
Since this task is triaged whenever issues are opened, but also when they are labeled, it can happen that this task will be triggered just when the "[Experiment] Automated labeling" label will be added. That means `getLabels()` will not include that label yet (because it is not yet part of the issue), but the new label will be the one that's added to the issue. This is what this logic is meant to catch. 

Does that make sense?

anomiex replied:
Why would `getLabels()` not include the label that was just added?

anomiex replied:
Looking through the history, I see you originally added code like this in #30100 in response to [this event](https://github.com/Automattic/wp-calypso/issues/75728#event-9000969272).

But there were a bunch of label manipulations there within the same minute. I suspect the re-addition was in a run triggered by [one of the earlier labels being added](https://github.com/Automattic/wp-calypso/issues/75728#event-9000957285) that saw one of the intermediate states rather than by the addition of `[Pri] Normal`.

jeherve replied:
You're right. I did a quick test, and the returned labels include the one that was just added:
https://github.com/jeherve/test-actions/issues/56

I pushed 398e8236e2339994de840afbb96a7bb557560a59 and it seems to work well, like in this example PR:
https://github.com/jeherve/jetpack/issues/154

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+	) {
+		return true;
+	}
+}
```
You say it returns boolean, but you're returning implicit-undefined here.

```suggestion

	return false;
}
```

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -114,6 +223,53 @@ async function triageIssues( payload, octokit ) {
+	// When an issue is first opened, parse its contents, send them to OpenAI,
+	// and add labels if any matching labels can be found.
+	// During testing, we'll run it for any issues, not just opened,
+	// bug only on issues with the "[Experiment] Automated labeling" label.
```
tyop 🙂
```suggestion
	// but only on issues with the "[Experiment] Automated labeling" label.
```

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/get-all-labels.js:
```diff
@@ -0,0 +1,43 @@
+ * @return {Promise<Array>} Promise resolving to an array of all labels in the repo.
+ */
+async function getAllLabels( octokit, owner, repo, filter = '' ) {
+	const labelList = [];
+	const cacheKey = `${ owner }/${ repo } #${ filter }`;
+	if ( cache[ cacheKey ] ) {
+		debug( `get-all-labels: Returning list of labels for ${ cacheKey } from cache.` );
+		return cache[ cacheKey ];
+	}
+
+	debug( `get-all-labels: Get list of labels for ${ cacheKey }.` );
+
+	for await ( const response of octokit.paginate.iterator( octokit.rest.issues.listLabelsForRepo, {
+		owner,
+		repo,
+		per_page: 100,
+	} ) ) {
+		for ( const label of response.data ) {
+			if ( filter && ! label.name.match( filter ) ) {
+				continue;
+			}
+			labelList.push( label );
+		}
+	}
+
+	cache[ cacheKey ] = labelList;
+	return labelList;
```
Probably more efficient to do it like this. The API call isn't changing based on the `filter` parameter.
```suggestion
	let labelList;
	const cacheKey = `${ owner }/${ repo }`;
	if ( cache[ cacheKey ] ) {
		debug( `get-all-labels: Using list of labels for ${ cacheKey } from cache.` );
		labelList = cache[ cacheKey ];
	} else {
		debug( `get-all-labels: Get list of labels for ${ cacheKey }.` );

		labelList = [];
		for await ( const response of octokit.paginate.iterator( octokit.rest.issues.listLabelsForRepo, {
			owner,
			repo,
			per_page: 100,
		} ) ) {
			for ( const label of response.data ) {
				labelList.push( label );
			}
		}
		cache[ cacheKey ] = labelList;
	}

	return filter ? labelList.filter( label => label.name.match( filter ) : labelList;
```

Or you might even drop the `filter` parameter entirely, and let the caller do `.filter( whatever )` if they want to filter the list.

jeherve replied:
done in f42ae3ac07652cd8d4ca06be8680729e6883789c

> Or you might even drop the filter parameter entirely, and let the caller do .filter( whatever ) if they want to filter the list.

I was thinking about that too, but I figure we may need it for other things in the future.

anomiex commented on projects/github-actions/repo-gardening/src/utils/openai/send-request.js:
```diff
@@ -0,0 +1,42 @@
+async function sendOpenAiRequest( message ) {
+	const apiKey = getInput( 'openai_api_key' );
+	if ( ! apiKey ) {
+		debug( 'openai: Input openai_api_key is required but missing. Aborting.' );
+		return '';
```
Totally optional, but this seems like something that should throw instead of just returning an empty string.

jeherve replied:
That's how I had set it up at first, but I figured it may not be ideal for folks outside of a8c who will not be able to use this OpenAI service.

anomiex replied:
I'd add a check for whether the key is set before calling this, or maybe a try-catch around the call.

jeherve replied:
What would you think about something like this?
d92472e2931093685cb957325a7e4622f46402db

anomiex replied:
Ok, left a few nitpicks inline.

anomiex commented on projects/github-actions/repo-gardening/src/utils/openai/send-request.js:
```diff
@@ -0,0 +1,42 @@
+
+		return completion?.choices?.[ 0 ]?.message?.content ?? '';
+	} catch ( error ) {
+		debug( `openai: Failed to send message to OpenAI. Error: ${ error }` );
+		return '';
```
This too.

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/get-all-labels.js:
```diff
@@ -0,0 +1,43 @@
+ * @param {GitHub} octokit - Initialized Octokit REST client.
+ * @param {string} owner   - Repository owner.
+ * @param {string} repo    - Repository name.
+ * @param {string} filter  - Optionally filter to only return a subset of labels. Use a regex pattern.
```
This seems more like what you intend.
```suggestion
 * @param {RegExp|string} filter  - Optionally filter to only return a subset of labels. Use a regex pattern.
```

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/get-all-labels.js:
```diff
@@ -0,0 +1,43 @@
+ * @param {string} filter  - Optionally filter to only return a subset of labels. Use a regex pattern.
+ * @return {Promise<Array>} Promise resolving to an array of all labels in the repo.
+ */
+async function getAllLabels( octokit, owner, repo, filter = '' ) {
```
I wonder whether "getAvailableLabels" might be more accurately descriptive.

jeherve replied:
I like it. f95209287d7735dfd08b4a149e836ba8e6f6c367

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+${ body }
+
+Here are the existing labels and their descriptions:
+${ repoLabels.join( ', ' ) }
```
This doesn't seem like it's going to do what you intend. As I read it, it'll do like this:

> Here are the existing labels and their descriptions:
> [object Object], [object Object], [object Object], [object Object], [object Object]

Seems like you might want to do something more like this?
```suggestion
${ repoLabels.map( label => '- ' + label.name + ': ' + label.description ).join( '\n' ) }
```
or maybe
```suggestion
${ repoLabels.map( label => '- ' + label.name + ': ' + ( label.description || '(no description)' ) ).join( '\n' ) }
```

jeherve replied:
Ah yes, I missed that! 🤦 . 1214685ea02db944fd541aaf4b429038e317bda8 should help.

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -11,6 +14,112 @@ const sendSlackMessage = require( '../../utils/slack/send-slack-message' );
+${ title }
+
+Here is the issue body:
+${ body }
```
I wonder whether the AI will get confused as to where the body ends and where the rest of the instructions begin.

jeherve replied:
That's a good point, I could improve that. Trying a05af0bf427313948ce3850e1850b99c912c407d, I will see if it helps.

anomiex reviewed (commented):
Gave it a once-over. As always, feel free to disagree with any of my suggestions.

jeherve committed:
Update projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/github-actions/repo-gardening/action.yml

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Update projects/github-actions/repo-gardening/src/utils/labels/get-all-labels.js

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Switch to a different model, gpt-4o-mini

From OpenAI:

> GPT-4o mini (“o” for “omni”) is our most advanced model in the small models category, and our cheapest model yet. It is multimodal (accepting text or image inputs and outputting text), has higher intelligence than gpt-3.5-turbo but is just as fast. It is meant to be used for smaller tasks, including vision tasks.
>
> We recommend choosing gpt-4o-mini where you would have previously used gpt-3.5-turbo as this model is more capable and cheaper.

jeherve committed:
Ensure labels are properly displayed

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739227591

jeherve committed:
Rename label getter

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739216852

anomiex committed:
Simplify label fetching / catching

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739211102

jeherve committed:
Adjust prompt to better separate each data point passed to OpenAI

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739230674

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into add/openai-auto-labeling-logic

jeherve committed:
Always display OpenAI response

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739184387

Co-authored-by: anomiex <anomiex@users.noreply.github.com>

jeherve committed:
Remove extra logging

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739191503

Co-authored-by: anomiex <anomiex@users.noreply.github.com>

jeherve committed:
Try requesting a different format back from OpenAI

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739193858

Co-authored-by: anomiex <anomiex@users.noreply.github.com>

jeherve committed:
Fail the job in case of OpenAI errors.

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739213017

jeherve committed:
Try a more strict response format

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739262852

OpenAI's Structured Outputs does not appear to be available for our model, gpt-4o-mini, yet.

jeherve committed:
Force a JSON Object response

jeherve committed:
Tweak prompt

jeherve committed:
Avoid shortcircuiting following logic when we get no results

jeherve committed:
Add a label marking processed entries

This is necessary during the experiment, since we trigger the process on multiple actions (opened, labeled, reopened). We don't want the labeling to be duplicated if an issue is opened AND labeled.

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -57,30 +56,28 @@ ${ body }
+The labels you will suggest must all come from the list below:
 
 ${ repoLabels
 	.map( label => `- ${ label.name }${ label?.description ? `: ${ label.description }` : '' }` )
```
With the new prompt here, could the AI decide the label is "[Feature] Something: This is actually the description" rather than the label being "[Feature] Something" with "This is actually the description" being a description of the label?

jeherve replied:
It hasn't seemed to be an issue so far, but we can try to be more precise in the prompt to ensure less potential issues:
0f58d58a2629aec378fe21d24e5d9a58d5d18595

anomiex replied:
I may be giving the AI too much or too little credit. 😀

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -57,30 +56,28 @@ ${ body }
 - Include 1 to 3 '[Feature]' labels.
 - Briefly explain each label choice in 1 sentence.
-- Format your response as a JSON object with 'labels' and 'explanations' keys.
+- Format your response as a JSON object, with each suggested label as a key, and its explanation as the value.
```
Would it help to be more explicit here?
```suggestion
- Format your response as a JSON object, with each suggested label as a key, and your explanation of the label choice as the value.
```

anomiex commented on projects/github-actions/repo-gardening/src/utils/openai/send-request.js:
```diff
@@ -1,19 +1,19 @@
+ * @param {string} responseFormat - Response format to use (plain by default, can be 'json_object').
  *
- * @return {Promise<string>} Promise resolving to the response from OpenAI.
+ * @return {Promise<string|void>} Promise resolving to the response from OpenAI, or void if an error occurred.
```
JS doesn't have a "void" type. It does have a "void" operator, but that produces the "undefined" value.
```suggestion
 * @return {Promise<string|undefined>} Promise resolving to the response from OpenAI, or undefined if an error occurred.
```

anomiex commented on projects/github-actions/repo-gardening/src/utils/openai/send-request.js:
```diff
@@ -1,19 +1,19 @@
 	if ( ! apiKey ) {
-		debug( 'openai: Input openai_api_key is required but missing. Aborting.' );
-		return '';
+		setFailed( 'openai: Input openai_api_key is required but missing.' );
```
Should there be a return here so it doesn't try to call OpenAI anyway?

jeherve replied:
Done in 46ff82fbabd0efa071199c60e4fdcec8f5e86d8a

jeherve committed:
Simplify label checking logic

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1739198113

jeherve committed:
customize prompt so it better understands the label list format.

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1746119187

jeherve committed:
Update return type

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Be more precise in expected json object response

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

anomiex committed:
Return early if no OpenAI key is set.

See https://github.com/Automattic/jetpack/pull/39152#discussion_r1746160510

anomiex reviewed (approved):
Looks reasonable, let's try it.

(I haven't tested it myself though)
